app/searchers/application_searcher/elasticsearch_query.rb

```diff
@@ -0,0 +1,57 @@
+class ApplicationSearcher
+  class KeywordHeuristics
+    def initialize clause
+      @clause = clause
+    end
+
+    def to_elasticsearch
+      return { exists: { field: @clause.field } } if @clause.blank?
+
+      return { term: { @clause.field => @clause.phrase.to_s } } if @clause.is_a? PhraseClause
+      { term: { @clause.field => @clause.term.to_s } }
+    end
+  end
+
+  class TextHeuristics
+    def initialize clause
+      @clause = clause
+    end
+
+    def to_elasticsearch
+      return { exists: { field: @clause.field } } if @clause.blank?
+
+      return { match_phrase: { @clause.field => @clause.phrase.to_s } } if @clause.is_a? PhraseClause
+      { match: { @clause.field => @clause.term.to_s } }
+    end
+  end
+
+  class ElasticsearchQuery
+    attr_reader :query
+
+    HEURISTICS_MAP = {
```

Comment:
@robacarp not super sure about this bit, but I can't find much of a cleaner way to do this all :/